crates/sui-storage/src/event_store/sql.rs

```diff
@@ -357,48 +368,57 @@ const QUERY_BY_OBJECT_ID: &str = "SELECT * FROM events WHERE timestamp >= ? AND
 impl EventStore for SqlEventStore {
     #[instrument(level = "debug", skip_all, err)]
     async fn add_events(&self, events: &[EventEnvelope]) -> Result<u64, SuiError> {
-        // TODO: submit writes in one transaction/batch so it won't just fail in the middle
         let mut cur_seq = self.seq_num.load(Ordering::Acquire);
         let initial_seq = cur_seq;
-        let mut rows_affected: u64 = 0;
-
-        // TODO: benchmark
-        // TODO: use techniques in https://docs.rs/sqlx-core/0.5.13/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values
-        // to execute all inserts in a single statement?
-        // TODO: See https://kerkour.com/high-performance-rust-with-sqlite
-        for event in events {
-            // Skip events that have a lower sequence number... which must be same or increasing
-            if event.seq_num < cur_seq {
-                debug!(tx_digest =? event.tx_digest, seq_num = event.seq_num, cur_seq, "Skipping event with lower sequence number than current");
+        let mut rows_affected = 0;
+
+        if events.is_empty() {
+            return Ok(0);
+        }
+
+        let mut start_index = 0;
+        let mut end_index = 0;
+        // This while loop only needed to make sure we don't write more than 1000 events at once.
+        // sqlx/SQLite has a limit of binding < 64k elements.
+        while end_index < events.len() {
+            if events[start_index].seq_num < cur_seq {
+                start_index += 1;
+                end_index += 1;
                 continue;
             }
-            cur_seq = event.seq_num;
-
-            // If batching, turn off persistent to avoid caching as we may fill up the prepared statement cache
-            let insert_tx_q = sqlx::query(SQL_INSERT_TX).persistent(true);
-            let event_type = EventType::from(&event.event);
-
-            let sender = event.event.sender().map(|sender| sender.to_vec());
-            let move_event_name = event.event.move_event_name();
-
-            // TODO: use batched API?
-            let res = insert_tx_q
-                .bind(event.timestamp as i64)
-                .bind(event.seq_num as i64)
-                .bind(event.tx_digest.map(|txd| txd.to_bytes()))
-                .bind(event_type as u16)
-                .bind(event.event.package_id().map(|pid| pid.to_vec()))
-                .bind(event.event.module_name())
-                .bind(event.event.object_id().map(|id| id.to_vec()))
-                .bind(Self::event_to_json(event))
-                .bind(move_event_name)
-                .bind(event.event.move_event_contents())
-                .bind(sender)
-                .bind(event.event.recipient_serialized()?)
+
+            end_index = (start_index + 1000).min(events.len());
+
```

Comment:
Should we also check the batch b/w start and end has events in monotonically increasing order? We seem to not be doing that anymore after this batched change. I wonder if we can still do that check?

Reply:
Hmmmm.   I guess I was hoping to skip that, it does increase the complexity.
If we find something out of order then we will have to skip it?